custom_metrics/sass.js

```diff
@@ -0,0 +1,309 @@
+	ret.variablesCombined = {};
+	scss.replace(/(?<=^|\s)--([\w-]+):\s*(.*#\{.*\$.+\}.*)\s*$/gm, ($0, name, value) => {
+		ret.variablesCombined["--" + name] = value;
+	});
+
+	// Heuristic for nesting &
+	ret.nested = {
+		"descendant": scss.match(/&\s+(?=[\w[.:#])/g)?.length,
+		">": scss.match(/&\s*(?=>)/g)?.length,
+		"+": scss.match(/&\s*(?=\+)/g)?.length,
+		"~": scss.match(/&\s*(?=~)/g)?.length,
+		"lone": scss.match(/&\s*(?=[,{]|$})/g)?.length,
+		"pseudo-element": scss.match(/&::/g)?.length,
+		"pseudo-class": scss.match(/&:(?!:)/g)?.length,
+		"id": scss.match(/&#\w+/g)?.length,
+		"class": scss.match(/&\.\w+/g)?.length,
+		"attr": scss.match(/&\[\w+/g)?.length,
+	};
+}
+
+let results = (async () => {
+
+const results = {};
+const $$ = s => [...document.querySelectorAll(s)];
+const sourcemapRegex = /\/\*[#@] sourceMappingURL=(.+?) \*\//;
+
+let stylesheets;
+
+if (typeof $WPT_BODIES !== "undefined") {
+	// in WPT, we already have the bodies
+	stylesheets = $WPT_BODIES.filter(request => request.type == "Stylesheet")
+		.map(file => {
+			return {url: file.url, body: file.response_body};
+		});
+}
+else {
+	// Outside WPT, use fetch()
+	stylesheets = await Promise.all($$("link[rel=stylesheet]").map(async link => {
+		try {
+			let response = await fetch(link.href);
+			let body = await response.text();
+
+			return {url: link.href, body};
+		}
+		catch (e) {}
+	}));
+}
+
+results.stylesheets = {
+	remote: stylesheets.length
+};
+
+stylesheets.push(...$$("style").map(s => {
+	return {
+		url: location,
+		body: s.textContent
+	}
+}));
+
+results.stylesheets.inline = stylesheets.length - results.stylesheets.remote;
+
+let sourcemapURLs = stylesheets.map(o => {
+	if (o) {
+		let url = o.body.match(sourcemapRegex)?.[1];
+
+		if (url) {
+			// Source map URL is relative to stylesheet URL
+			return new URL(url, o.url);
+		}
+	}
+}).filter(url => !!url);
+
+results.sourcemaps = {
+	count: sourcemapURLs.length,
+	ext: {}
+};
+
+if (sourcemapURLs.length === 0) {
+	return results;
+}
+// console.log("sourcemap URLs", sourcemapURLs);
+
+// Assumption: Either all sources are SCSS or none.
+let scss = await Promise.all(sourcemapURLs.map(async url => {
+	try {
+		var response = await fetch(url);
+		var json = await response.json();
+	}
+	catch (e) {
+		return;
+	}
+
+	let sources = json.sources;
+	let base = json.sourceRoot? new URL(json.sourceRoot, url) : url;
+	let scss = 0;
+
+	sources = sources.map(s => {
+		let url = new URL(s, base);
+		let ext = url.pathname.match(/\.(\w+)$/)?.[1];
+
+		if (ext) {
+			results.sourcemaps.ext[ext] = (results.sourcemaps.ext[ext] || 0) + 1;
+
+			if (ext === "scss") {
+				scss++;
+			}
+		}
+
+		return url;
+	});
+
+	if (scss > 0) {
+		if (json.sourcesContent) {
+			// Source is already here, no more requests needed, yay!
+			return json.sourcesContent;
+		}
+
+		let code = await Promise.all(sources.map(async s => {
+			let response = await fetch(s);
+			let text = response.ok? await response.text() : "";
```

Comment:
🤯

Reply:
??

Reply:
lol nevermind, I assumed the `?` was some new usage of nullish coalescing but upon closer inspection it's just a ternary operator (it was late 😅)